tests/check_docs.mts

```diff
@@ -0,0 +1,130 @@
+/**
+ * Require that documentation is kept up to date as the referenced files change.
+ *
+ * To do this, for each file in the docs directory, we look at the set of files
+ * listed in the front matter and check when each of them last changed. If any
+ * of them is more recent than either the last commit to the doc file or the
+ * commit listed in the "updated" field (whichever is more recent), then the doc
+ * is out of date.
+ */
+import child from 'node:child_process';
+import { promises as fs } from 'node:fs';
+import path from 'node:path';
+import { fileURLToPath } from 'node:url';
+import { promisify } from 'node:util';
+import matter from 'gray-matter';
+
+const execFile = promisify(child.execFile);
+
+const filename = fileURLToPath(import.meta.url);
+const dirname = path.dirname(filename);
+
+interface NewerFile {
+  file: string;
+  commit: string;
+}
+
+type DocError = {
+  type: 'out-of-date'
+  doc: string;
+  newerFiles: NewerFile[];
+} | {
+  type: 'missing-front-matter';
+  doc: string;
+} | {
+  type: 'missing-updated-field';
+  doc: string;
+} | {
+  type: 'missing-files';
+  doc: string;
+}
+
+const checkDoc = async (doc: string): Promise<DocError | undefined> => {
+  const docMatter = matter(await fs.readFile(doc));
+
+  // First: figure out how recent the documentation is. This can either be the
+  // "updated" field in the front matter, or, if the last commit to the file
+  // contains the updated commit, the last commit.
+  let docRevision: string = docMatter.data.updated;
+  if (!docRevision || typeof docRevision !== 'string') {
+    return {
+      type: 'missing-updated-field',
+      doc,
+    };
+  }
+
+  const { stdout: fileRevision } = await execFile('git', ['rev-list', '-1', 'HEAD', `^${docRevision}`, '--', doc]);
+  if (fileRevision.trim().length > 0) {
+    docRevision = fileRevision.trim();
+  }
```

Comment:
For the record, this logic is kind of weird, and I'm not sure if it's right. The commit in the `updated` frontmatter field is always going to be less recent than the most recent commit to the file itself, because :sparkles:merkle trees:sparkles:.

I do want it to be possible to flag "this documentation file wasn't changed but is still up to date", but I also would prefer to not have, like, a "generation number" in the documentation file or whatever that just gets monotonically incremented. A commit ID at least feels like it has _some_ semantic meaning.

Reply:
A commit ID has some meaning, but yeah, it definitely feels less than ergonomic, especially if the linter isn't going to tell you what the actual commit ID you should be updating the file to is.  Maybe we should have a helper script `mark-up-to-date` that does the appropriate git lookup & find/replace in the front matter?  or outputs a diff that you can pipe into `patch` or something?

Something I've seen done in other documentation systems is to have a "last-reviewed-at" date, but that's typically more oriented for process/team-level documentation rather than doc documentation (and skew between authored/committed dates during code review could be troublesome) and is more targeted at an async "audit any docs that haven't been updated in 3 months to make sure they're still accurate" flow rather than a blocking "if you touched the code, you must also touch the docs with a fresher sha1".

A thing that I've found tricky in the past is when the logic for something moves between locations, but the docs still reference the old home for that logic.  I guess if we check for changes in all the files referenced, then any time you extract logic from one of them you're modifying that one, so your scheme covers that case.  I suspect this may wind up adding a chunk of friction to refactorings, though?  I don't really have a great answer for how to get around that, though; it seems intrinsic to making the docs know about what files their content covers.

To be honest, while I certainly value having up-to-date docs, I'm cautious about making doc-freshness checks something that blocks merges.  One possible negative effect would be that it may incentivize us to bump shas blindly without actually doing meaningful review of the docs, but on the flipside: they'd probably force us to at least see them/be reminded they exist in code review, and if semantics are indeed changing, that'd be a useful prompt, and also we're generally under less schedule pressure than the other places I've seen this sort of thing break down.

Reply:
For the record, I never followed up on this because you correctly called out a thing that I was already unhappy with and I didn't have a better option. However, helping @egnor with getting JR setup has reminded me that I want to do more here (especially because _I've_ forgotten how some of these systems fit together, unsurprisingly), so I'm going to re-raise it.

I mostly agree that this type of enforcement system can be problematic, but I think it's more problematic to not have anything at all and allow the docs to drift. I'm curious to see what the actual false-negative rate on this test is, but my gut says that it's going to be relatively low. The exception will be large multi-purpose files that lead to unrelated changes (e.g. heaven forbid we document anything in `PuzzlePage.tsx`), but I think that's a good motivation to split out those files and make them more manageable?

I am unhappy with the git revision thing, though, for the reasons I mentioned above. But maybe a "last updated date" would actually work. With a last updated date, we don't actually need to look at the value of the date, we just look and see if the docs are at-least-as-fresh as the files they document.

Reply:
OK. Reworked this so that `updated` is now a date, and the actual value is ignored by the linter. If you update files that are referenced by documentation but don't need to update the documentation itself, you can bump the `updated` timestamp, which will cause the file to appear in the git history as being updated at the same time or later than the files it documents. I'm happier with this.